### DISCUSSION TITLE
Slower when computing across nodes

### DISCUSSION THREAD

**Discussion by stonebeta** [idaholab/moose#19598] [Q&A General] [answered]
Hello everyone.
I find that it is slower when computing across nodes.
This is my input file: [input.zip](https://github.com/idaholab/moose/files/7691793/input.zip) 
It is basically the same as [part_1.2.i](https://github.com/idaholab/moose/blob/6ad3700a7a2265d14bbd51e64232b35217d3b1fe/modules/tensor_mechanics/tutorials/basics/part_1.2.i).
There are some changes in the Mesh, BCs and Executioner:
```
[Mesh]
  file = necking_quad4.e
  uniform_refine = 6
[]
[BCs]
  [./left]
    type = DirichletBC
    variable = disp_x
    boundary = left
    value = 0.0
  [../]
  [./bottom]
    type = DirichletBC
    variable = disp_y
    boundary = bottom
    value = 0.0
  [../]
  [./top]
    type = FunctionDirichletBC
    variable = disp_y
    boundary = top
    function = 't'
  [../]
[]
[Executioner]
  type = Transient
  dt = 0.001
  end_time = 0.6
  solve_type = 'NEWTON'
  petsc_options = '-snes_ksp_ew'
  petsc_options_iname = '-pc_type -sub_pc_type -pc_asm_overlap -ksp_gmres_restart'
  petsc_options_value = 'asm lu 1 101'
[]
```
26 processors, 52 processors and 104 processors (2 nodes, 52 for each) were used. 
The following is their performance log:
![ti](https://user-images.githubusercontent.com/85184413/145554951-dfd89b0e-04ae-44ed-9b77-9547c9aba878.png)
And their full logs: [log.zip](https://github.com/idaholab/moose/files/7691853/log.zip)
 
I find that the time of `FEProblem::computeResidualInternal` and `FEProblem::computeJacobianInternal` in the Total(s) column is almost doubled when computing across nodes. I guess something's wrong. Please help me.

**Comment by GiudGiud** [comment, marked as answer]
Hello

There are many things to be mindful about when doing parallel simulations to get good scaling far out. 

First you should try to use a distributed mesh (--distributed-mesh) so that mesh operations are not replicated (here, you are spending most of the time computing the residual, so it should not be the problem)

Then you should use a preconditioner that scales well. SMP+asm should be ok. I dont know if it s the best fitted for this particular physics. 
How are the number of iterations evolving as the number of processes is increased?

You should make sure you have enough DOFs on every rank. At least 20,000. What does it say in the beginning of the log?
Uniform refine = 6 is quite a bit of refinement, but still good to check. 

Guillaume
PS: I would also look through other discussions to find more advice previously given on this forum by other devs.

**Reply by stonebeta** [reply]
Thank you for your help. I should use the distributed mesh.
Here is the result using the distributed mesh. The end_time was changed from 0.6 to 0.3.
![test2](https://user-images.githubusercontent.com/85184413/145677778-292a4c0e-73e2-4db3-8887-67616606451b.png)
It looks much better. Thank you.

**Comment by hugary1995** [comment]
Comunnication between nodes can have a.fairly large overhead depending on the architecture of your cluster. I observe the same thing on some of the nodes on our university's cluster, if there is a way to improve the speed, I'm all ears.
